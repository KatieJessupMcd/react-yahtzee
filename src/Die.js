import React, { Component } from 'react';
import './Die.css';

class Die extends Component {
  render() {
    return (
      <button
        className={this.props.locked ? 'Die Die-locked' : 'Die'}
        // BUG arrow function to maintian this
        onClick={() => this.props.handleClick(this.props.idx)}
      >
        {this.props.val}
      </button>
    );
  }
}

export default Die;
