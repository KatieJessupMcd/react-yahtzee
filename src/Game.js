import React, { Component } from 'react';
import Dice from './Dice';
import ScoreTable from './ScoreTable';
import './Game.css';

const NUM_DICE = 5;
const NUM_ROLLS = 3;

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dice: Array.from({ length: NUM_DICE }),
      locked: Array(NUM_DICE).fill(false),
      rollsLeft: NUM_ROLLS,
      disabled: false,
      scores: {
        ones: undefined,
        twos: undefined,
        threes: undefined,
        fours: undefined,
        fives: undefined,
        sixes: undefined,
        threeOfKind: undefined,
        fourOfKind: undefined,
        fullHouse: undefined,
        smallStraight: undefined,
        largeStraight: undefined,
        yahtzee: undefined,
        chance: undefined
      }
    };
    this.roll = this.roll.bind(this);
    this.doScore = this.doScore.bind(this);
    // BUG - this was not bound before
    this.toggleLocked = this.toggleLocked.bind(this);
    this.disableScoreRow = this.disableScoreRow.bind(this);
  }

  roll(evt) {
    // roll dice whose indexes are in reroll
    this.setState(st => ({
      dice: st.dice.map((d, i) =>
        st.locked[i] ? d : Math.ceil(Math.random() * 6)
      ),
      locked: st.rollsLeft > 1 ? st.locked : Array(NUM_DICE).fill(true),
      disabled: st.rollsLeft > 1 ? false : true,
      rollsLeft: st.rollsLeft - 1
    }));
  }

  toggleLocked(idx) {
    // toggle whether idx is in locked or not
    this.setState(st => ({
      locked: [
        ...st.locked.slice(0, idx),
        !st.locked[idx],
        ...st.locked.slice(idx + 1)
      ]
    }));
  }

  doScore(rulename, ruleFn) {
    // evaluate this ruleFn with the dice and score this rulename
    this.setState(st => ({
      scores: { ...st.scores, [rulename]: ruleFn(this.state.dice) },
      rollsLeft: NUM_ROLLS,
      locked: Array(NUM_DICE).fill(false)
    }));
    this.disableScoreRow(rulename);
    this.roll();
  }

  disableScoreRow(name) {
    console.log(this.state);
    this.setState({
      disabled: true
    });
  }

  render() {
    return (
      <section>
        <Dice
          dice={this.state.dice}
          locked={this.state.locked}
          handleClick={this.toggleLocked}
          // disabled is going to have a TRUE OR FALSE;
          disabled={this.state.disabled}
        />
        <button
          className="Game-reroll"
          disabled={this.state.locked.every(x => x)}
          onClick={this.roll}
        >
          {this.state.rollsLeft} Rerolls Left
        </button>
        <ScoreTable
          // disabled={this.state.disabled}
          doScore={this.doScore}
          scores={this.state.scores}
        />
      </section>
    );
  }
}

export default Game;
