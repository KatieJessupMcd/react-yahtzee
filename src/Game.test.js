import React from 'react';
import ReactDOM from 'react-dom';
import Game from './Game';
import Dice from './Dice';
import Die from './Die';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

it('renders without crashing', function() {
  shallow(<Game />);
});
// snapshot test
it('matches snapshot', function() {
  let wrapper = shallow(<Game />);
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});

it('should toggle a locked arr value to True on first die button click', function() {
  let wrapper = mount(<Game />);

  expect(wrapper.state().locked[0]).toBe(false);

  const dieButton = wrapper.find(Die).at(0);
  dieButton.simulate('click');
  let wrapperState = wrapper.state();

  expect(wrapperState.locked[0]).toBe(true);
});
