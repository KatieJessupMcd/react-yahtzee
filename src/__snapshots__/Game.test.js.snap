// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section>
  <Dice
    dice={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
    handleClick={[Function]}
    locked={
      Array [
        false,
        false,
        false,
        false,
        false,
      ]
    }
  />
  <button
    className="Game-reroll"
    disabled={false}
    onClick={[Function]}
  >
    3
     Rerolls Left
  </button>
  <ScoreTable
    doScore={[Function]}
    scores={
      Object {
        "chance": undefined,
        "fives": undefined,
        "fourOfKind": undefined,
        "fours": undefined,
        "fullHouse": undefined,
        "largeStraight": undefined,
        "ones": undefined,
        "sixes": undefined,
        "smallStraight": undefined,
        "threeOfKind": undefined,
        "threes": undefined,
        "twos": undefined,
        "yahtzee": undefined,
      }
    }
  />
</section>
`;

exports[`should toggle a locked arr value to True on die button click 1`] = `
<section>
  <Dice
    dice={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
    handleClick={[Function]}
    locked={
      Array [
        false,
        false,
        false,
        false,
        false,
      ]
    }
  />
  <button
    className="Game-reroll"
    disabled={false}
    onClick={[Function]}
  >
    3
     Rerolls Left
  </button>
  <ScoreTable
    doScore={[Function]}
    scores={
      Object {
        "chance": undefined,
        "fives": undefined,
        "fourOfKind": undefined,
        "fours": undefined,
        "fullHouse": undefined,
        "largeStraight": undefined,
        "ones": undefined,
        "sixes": undefined,
        "smallStraight": undefined,
        "threeOfKind": undefined,
        "threes": undefined,
        "twos": undefined,
        "yahtzee": undefined,
      }
    }
  />
</section>
`;
